Extension type with generic

It is impossible to use the library if an extension type contains generics.
For extension types without generics, it works fine (at the master branch only).

- pubspec.yaml
```yml
dev_dependencies:
  # https://github.com/dart-lang/mockito/issues/732
  # https://github.com/dart-lang/mockito/pull/738
  # https://github.com/dart-lang/build/issues/4495
  mockito:
    git:
      url: https://github.com/dart-lang/mockito.git
      ref: master
```

- example.dart
```dart
import 'dart:async';

import 'package:mockito/annotations.dart';

extension type ET$Generic<T extends num>(Stream<T> _) implements Stream<T> {
  ET$Generic<S> filter<S extends num>() =>
      ET$Generic<S>(transform<S>(StreamTransformer<T, S>.fromHandlers(
        handleData: (data, sink) => switch (data) {
          S valid => sink.add(valid),
          _ => null,
        },
      )));
}

abstract interface class Interface {
  abstract final ET$Generic<num> stream;
}

@GenerateMocks([Interface])
void main() {}
```

- output
```log
Built build_runner:build_runner.
[INFO] Generating build script completed, took 129ms
[INFO] Reading cached asset graph completed, took 46ms
[INFO] Checking for updates since last build completed, took 328ms
[SEVERE] mockito:mockBuilder on test/unit/example.dart:

Bad state: T extends num not found, scopes: [{}]
package:mockito/src/builder.dart 2024:8                      _MockClassInfo._lookupTypeParameter
package:mockito/src/builder.dart 2162:22                     _MockClassInfo._typeReference.<fn>
package:code_builder/src/specs/type_reference.g.dart 164:33  _$TypeReferenceBuilder.update
package:code_builder/src/specs/type_reference.g.dart 22:36   new _$TypeReference
package:mockito/src/builder.dart 2160:14                     _MockClassInfo._typeReference
package:mockito/src/builder.dart 1527:27                     _MockClassInfo._dummyValue
package:mockito/src/builder.dart 1650:11                     _MockClassInfo._dummyValueImplementing
package:mockito/src/builder.dart 1547:12                     _MockClassInfo._dummyValue
package:mockito/src/builder.dart 1945:24                     _MockClassInfo._buildOverridingGetter
package:mockito/src/builder.dart 1222:36                     _MockClassInfo.fieldOverrides.<fn>
package:code_builder/src/specs/method.g.dart 323:33          _$MethodBuilder.update
package:code_builder/src/specs/method.g.dart 38:29           new _$Method
package:mockito/src/builder.dart 1222:15                     _MockClassInfo.fieldOverrides
dart:core                                                    List.addAll
package:built_collection/src/list/list_builder.dart 98:14    ListBuilder.addAll
package:mockito/src/builder.dart 1180:28                     _MockClassInfo._buildMockClass.<fn>.<fn>
package:mockito/src/builder.dart 2069:24                     _MockClassInfo._withTypeParameters
package:mockito/src/builder.dart 1134:7                      _MockClassInfo._buildMockClass.<fn>
package:code_builder/src/specs/class.g.dart 345:33           _$ClassBuilder.update
package:code_builder/src/specs/class.g.dart 40:28            new _$Class
package:mockito/src/builder.dart 1112:12                     _MockClassInfo._buildMockClass
package:mockito/src/builder.dart 1058:9                      new _MockLibraryInfo
package:mockito/src/builder.dart 94:29                       MockBuilder.build

[INFO] Running build completed, took 384ms
[INFO] Caching finalized dependency graph completed, took 32ms
[SEVERE] Failed after 421ms
make: *** [generate] Error 1
```

--- a/builder_pkgs/mockito/CHANGELOG.md
+++ b/builder_pkgs/mockito/CHANGELOG.md
@@ -1,13 +1,17 @@
 ## 5.8.1-wip
 
+* Cast dummy values for extension types to the extension type so they satisfy
+  static type checking in generic return types like `Future<ExtensionType>`.
+  [#4590](https://github.com/dart-lang/build/issues/4590)
+* Use canonical asset path in generated header comments so symlink paths are
+  excluded.
+  [#4545](https://github.com/dart-lang/build/issues/4545)
 * Deterministically select library to import when multiple libraries export the
   same type.
   [#4549](https://github.com/dart-lang/build/issues/4549)
 * Prefer importing public entrypoints over `/src/` implementation libraries
   when multiple libraries export the same type.
   [#4551](https://github.com/dart-lang/build/issues/4551)
-  excluded.
-  [#4545](https://github.com/dart-lang/build/issues/4545)
 * Throw `InvalidMockitoAnnotationException` with a clear message when
   `@GenerateMocks` `customMocks` or `@GenerateNiceMocks` `mocks` is not a list
   literal, instead of crashing with a type cast exception.

--- a/builder_pkgs/mockito/lib/src/builder.dart
+++ b/builder_pkgs/mockito/lib/src/builder.dart
@@ -2227,7 +2227,7 @@ class _MockClassInfo {
     _dummyFakedValue(dartType, invocation),
     ExtensionTypeElement(:final typeErasure)
         when !typeErasure.containsPrivateName =>
-      _dummyValue(typeErasure, invocation),
+      _dummyValue(typeErasure, invocation).asA(_typeReference(dartType)),
     ExtensionTypeElement() => _dummyValueFallbackToRuntime(
       dartType,
       invocation,

--- a/builder_pkgs/mockito/test/builder/auto_mocks_test.dart
+++ b/builder_pkgs/mockito/test/builder/auto_mocks_test.dart
@@ -4255,7 +4255,24 @@ void main() {
           E get v;
         }
         '''),
-        decodedMatches(allOf(contains('E get v'), contains('returnValue: 0'))),
+        decodedMatches(
+          allOf(contains('E get v'), contains('returnValue: (0 as _i2.E)')),
+        ),
+      );
+    });
+
+    test('are supported as type arguments in return types', () async {
+      await expectSingleNonNullableOutput(
+        dedent('''
+        extension type E(int v) {}
+        class Foo {
+          Future<E> get v;
+        }
+        '''),
+        _containsAllOf(
+          'Future<_i2.E> get v',
+          'returnValue: _i3.Future<_i2.E>.value((0 as _i2.E))',
+        ),
       );
     });
   });

--- a/builder_pkgs/mockito/test/end2end/generated_mocks_test.mocks.dart
+++ b/builder_pkgs/mockito/test/end2end/generated_mocks_test.mocks.dart
@@ -1002,8 +1002,8 @@ class MockUsesExtTypes extends _i1.Mock implements _i3.UsesExtTypes {
   _i3.Ext extTypeReturn(int? _0) =>
       (super.noSuchMethod(
             Invocation.method(#extTypeReturn, [_0]),
-            returnValue: 0,
-            returnValueForMissingStub: 0,
+            returnValue: (0 as _i3.Ext),
+            returnValueForMissingStub: (0 as _i3.Ext),
           )
           as _i3.Ext);
 

